service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -1,93 +1,119 @@
 package com.epam.esm.service.impl;
 
 import com.epam.esm.dao.CertificateDao;
-import com.epam.esm.dao.parameters.ParametersProvider;
-import com.epam.esm.dao.parameters.SearchQueryProvider;
+import com.epam.esm.dao.TagDao;
+import com.epam.esm.service.exception.CreateEntityInternalException;
+import com.epam.esm.service.exception.UpdateEntityInternalException;
 import com.epam.esm.domain.dto.CertificateDTO;
-import com.epam.esm.domain.dto.TagDTO;
+import com.epam.esm.domain.util.SearchParameter;
 import com.epam.esm.domain.entity.Certificate;
 import com.epam.esm.domain.entity.Tag;
 import com.epam.esm.service.CertificateService;
-import com.epam.esm.service.TagService;
 import com.epam.esm.service.exception.IdNotExistException;
-import com.epam.esm.service.mapper.CertificateDTOMapper;
+import org.modelmapper.ModelMapper;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
 import java.util.List;
+import java.util.Objects;
 import java.util.stream.Collectors;
 
 @Service
 public class CertificateServiceImpl implements CertificateService {
 
     private final CertificateDao certificateDAO;
-    private final CertificateDTOMapper certificateDTOMapper;
-    private final TagService tagService;
-    private final SearchQueryProvider queryProvider;
-    private final ParametersProvider queryParametersProvider;
-
-    public CertificateServiceImpl(CertificateDao certificateDAO, CertificateDTOMapper certificateDTOMapper,
-                                  TagService tagService, SearchQueryProvider queryProvider,
-                                  ParametersProvider queryParametersProvider) {
+    private final TagDao tagDao;
+    private final ModelMapper modelMapper;
+
+    public CertificateServiceImpl(CertificateDao certificateDAO, TagDao tagDao, ModelMapper modelMapper) {
         this.certificateDAO = certificateDAO;
-        this.certificateDTOMapper = certificateDTOMapper;
-        this.tagService = tagService;
-        this.queryProvider = queryProvider;
-        this.queryParametersProvider = queryParametersProvider;
+        this.tagDao = tagDao;
+        this.modelMapper = modelMapper;
     }
 
     @Override
     @Transactional
-    public Integer update(CertificateDTO certificate) {
-        return certificateDAO.update(certificateDTOMapper.toEntity(certificate));
+    public CertificateDTO update(CertificateDTO dto) {
+        Certificate existing = certificateDAO.readById(dto.getId())
+                .orElseThrow(() -> new IdNotExistException(dto.getId().toString()));

```

Comment:
This validation already implemented in *readById* method.

Reply:
thank you, fixed